### PR TITLE
WIP: Ws client code

### PR DESCRIPTION
**Pull request by gideonthomas**

### CODE REVIEW

File: client/src/index.js [hunk 1]
```diff
@@ -153,7 +154,7 @@ function createFS() {
 
       syncInterval = setInterval(function() {
         if(needsSync) {
-          sync.request();
+          sync.request('/');
```
Comment (alicoding) [inline client/src/index.js:156]: Is this mean root directory?

File: client/src/index.js [hunk 2]
```diff
@@ -52,7 +52,7 @@
  * sync.SYNC_ERROR = 4
  */
 
-var MakeDriveSync = require('./interface');
+var MakeDriveSync = require('./sync');
```
Comment (humphd) [inline client/src/index.js:55]: We have so many things called 'sync`, I wonder if we should consider a better name here.

File: client/src/index.js [hunk 3]
```diff
@@ -63,6 +63,7 @@ module.exports = MakeDrive;
 // tabs/windows may also be using this same instance (i.e., fs
 // is shared at the provider level).
 var _fs;
+var needsSync = false;
```
Comment (humphd) [inline client/src/index.js]: Why this change?  It doesn't need to be global.

File: client/src/index.js (hunk 1 above)
Comment (humphd) [inline client/src/index.js:156]: This isn't strictly necessary (i.e., the `request` function will default to `'/'` if path is omitted or doesn't exist.  However, I think it's good to be explicit in this case.  Let's do this.

File: client/src/sync.js [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+};
+
+var syncState,
+    syncStep;
+
+function callback(err) {
+  if(syncCallback) {
+    var callbackFunction = syncCallback;
+    syncCallback = null;
+    return callbackFunction(err);
```
Comment (humphd) [inline client/src/sync.js]: Modify this function like so:

``` javascript
function maybeCallback(err) {
  if(syncCallback) {
    syncCallback(err);
    syncCallback = null;
  }
}
```

Then you can call `maybeCallback(...)` below when necessary.

File: client/src/sync.js [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+function handleMessage(data, flags) {
+  data = data.data;
+  try {
+    data = JSON.parse( data );
+  } catch( e ) {
+    sync.state = sync.SYNC_ERROR;
+    return sync.emit('error', e);
+  }
+  // Is the object a response?
```
Comment (humphd) [inline client/src/sync.js]: Newline here (after 46).

File: client/src/sync.js [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+  if (data.type === SyncMessage.REQUEST) {
```
Comment (humphd) [inline client/src/sync.js]: You do too much in each function.  Break things up.  The pattern should be:

``` javascript
// Is the object a response?
if(data.type === SyncMessage.REQUEST) {
  handleRequestMessage(...);
} else if (data.type === SyncMessage.RESPONSE) {
  handleResponseMessage(...);
} else if (data.type === SyncMessage.ERROR) {
  handleErrorMessage(...);
}
```

Then you can move things into separate functions.  Try to separate control flow like this (i.e., choosing paths based on message types) from actual work you want to do once you have the correct path.  It makes understanding the code, and thus maintaining it, a lot easier.

File: client/src/sync.js [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+          return sync.emit('error', err);
+        }
+        sync.state = sync.SYNC_SYNCING;
+        sync.emit('syncing');
+        syncStep = steps.PATCH;
+
+        var message = new SyncMessage(SyncMessage.REQUEST, SyncMessage.DIFFS);
+        message.setContent({checksums: checksums});
+
+        return socket.send(JSON.stringify(message));
```
Comment (humphd) [inline client/src/sync.js]: You don't need a `return` here.

File: client/src/sync.js [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+      var srcList = data.content.srcList;
+      path = data.content.path;
+
+      return rsync.checksums(fs, path, srcList, rsyncOptions, function(err, checksums) {
+        if (err) {
+          syncStep = steps.FAILED;
+          sync.state = sync.SYNC_ERROR;
+          return sync.emit('error', err);
+        }
+        sync.state = sync.SYNC_SYNCING;
```
Comment (humphd) [inline client/src/sync.js]: Isn't this late to be doing this?  Shouldn't you change the state before you start the checksums?  I'd put this state/event code at 53.

File: client/src/sync.js [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+        if(err){
+          syncState = states.ERROR;
+          return callback(err);
+        }
+        syncStep = steps.PATCH;
+        var message = new SyncMessage(SyncMessage.RESPONSE, SyncMessage.DIFFS);
+        diffs = serializeDiff(diffs);
+        message.setContent({diffs: diffs});
+
+        return socket.send(JSON.stringify(message));
```
Comment (humphd) [inline client/src/sync.js]: No `return` here.

File: client/src/sync.js [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+      });
+    }
+    sync.state = sync.SYNC_ERROR;
```
Comment (humphd) [inline client/src/sync.js]: Why are we in an error state here?  Can you add comments above this?

File: client/src/sync.js [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+      return rsync.sourceList(fs, path, rsyncOptions, function(err, srcList){
+        if(err){
+          syncState = states.ERROR;
+          return callback(err);
+        }
+        syncStep = steps.DIFFS;
+        var message = new SyncMessage(SyncMessage.REQUEST, SyncMessage.CHKSUM);
+        message.setContent({srcList: srcList, path: path});
+
+        return socket.send(JSON.stringify(message));
```
Comment (humphd) [inline client/src/sync.js]: You don't need `return` here.

File: client/src/sync.js [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+      });
+    }
+    // UPSTREAM - PATCH
+    if(data.name === SyncMessage.PATCH && syncstate === states.SYNCING && syncStep === steps.PATCH){
+      syncState = states.READY;
+      syncStep = steps.SYNCED;
+      sync.state = sync.SYNC_CONNECTED;
+      return callback();
```
Comment (humphd) [inline client/src/sync.js:39]: For future, tell yourself that when you have to use this many `return` statements, you should actually be breaking things up into separate functions.  This is a sign that you've got too many paths joined into a single function.  In reality, each of your `return` statements marks the boundary of a sub-process, which should be placed into a separate function.

File: client/src/sync.js [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+      diffs = deserializeDiff(diffs);
+      return rsync.patch(fs, path, diffs, rsyncOptions, function(err) {
+        if (err) {
+          syncStep = steps.FAILED;
+          sync.state = sync.SYNC_ERROR;
+          return sync.emit('error', err);
+        }
+        syncStep = steps.SYNCED;
+        sync.state = sync.SYNC_CONNECTED;
+        return sync.emit('completed');
```
Comment (humphd) [inline client/src/sync.js]: No `return` here.

File: client/src/sync.js [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+    if(((data.name === SyncMessage.SRCLIST && syncStep === steps.SYNCED) || (data.name === SyncMessage.DIFFS && syncStep === steps.SYNCED)) && syncState === states.READY){
+      syncStep = steps.FAILED;
+      sync.state = sync.SYNC_ERROR;
+      //TODO: handle what to do to reinitiate downstream sync
+      return sync.emit('error', new Error('Could not sync filesystem from server'));
+    }
+    // UPSTREAM - LOCK
+    if(data.name === SyncMessage.LOCKED && syncState === states.READY && syncStep === steps.SYNCED){
+      syncState = states.ERROR;
+      return callback(new Error('Cannot sync at the moment. Try again later!'));
```
Comment (humphd) [inline client/src/sync.js]: It's more than just that you "cannot sync at the moment"--it's really that there is a lock.  Be more clear.

File: client/src/sync.js [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+    }
+    // UPSTREAM - ERROR
+    if(((data.name === SyncMessage.CHKSUM && syncStep === steps.DIFFS) || (data.name === SyncMessage.PATCH && syncStep === steps.PATCH)) && syncState === states.SYNCING){
+      syncState = states.ERROR;
+      syncStep = steps.FAILED;
+      return callback(new Error('Fatal error: Failed to sync to server'));
+    }
+    return sync.emit('error', data.content.error);
```
Comment (humphd) [inline client/src/sync.js]: Add a comment here to describe why you have a catch all.

File: client/src/sync.js [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+  }
+  sync.state = sync.SYNC_ERROR;
+  return sync.emit('error', new Error('Cannot handle message'));
```
Comment (humphd) [inline client/src/sync.js]: No `return` here.

File: client/src/sync.js [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+function handleAuth(data, flags) {
+  data = data.data;
+  try {
+    data = JSON.parse(data);
+  } catch(e) {
+    return callback(e);
+  }
+  if(data.token) {
```
Comment (humphd) [inline client/src/sync.js]: Add newline before this.

File: client/src/sync.js [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+function handleClose(code, data) {
+  data = data.data;
+  try {
+    data = JSON.parse(data);
+  } catch(e) {
+    sync.emit('error', e);
+    sync.state = sync.SYNC_DISCONNECTED;
+    return sync.emit('disconnected');
+  }
+  socket.close();
```
Comment (humphd) [inline client/src/sync.js]: Newline before this.

File: client/src/sync.js [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+  var error = new Error(code + ': ' + data);
+  sync.emit('error', error);
+  sync.state = sync.SYNC_DISCONNECTED;
+  return sync.emit('disconnected');
```
Comment (humphd) [inline client/src/sync.js]: No `return` here.

File: client/src/sync.js [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+function init(sync, url, fileSystem, callbackFunction) {
+  syncInterface = sync;
+  syncCallback = callbackFunction;
+  fs = fileSystem;
+  socket = new WebSocket(url);
+  socket.onmessage = handleAuth;
+  socket.onclose = handleClose;
```
Comment (humphd) [inline client/src/sync.js:62]: Is there not an `onerror` handler you should also register here?

File: client/src/sync.js [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+  }
+  socket.close();
+  var error = new Error(code + ': ' + data);
+  sync.emit('error', error);
+  sync.state = sync.SYNC_DISCONNECTED;
+  return sync.emit('disconnected');
+}
+
+function init(sync, url, fileSystem, callbackFunction) {
+  syncInterface = sync;
```
Comment (humphd) [inline client/src/sync.js]: I don't love the pattern of globals you use throughout this file.  I think it might make more sense for each sync to be an instance object, where `init()` calls a ctor, and then you can call a method `sync` on that instance, which does all the steps above.  Then the callback, path, sync, filesystem, etc can be properties or locals on that instance.
